Client/Wallet/SolanaProviderHelper.swift

```diff
@@ -0,0 +1,336 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import WebKit
+import BraveCore
+import BraveShared
+import struct Shared.Logger
+
+private let log = Logger.browserLogger
+
+class SolanaProviderHelper: TabContentScript {
+  
+  fileprivate enum Keys: String {
+    case connect
+    case disconnect
+    case signMessage
+    case message
+    case serializedMessage
+    case signatures
+    case sendOptions
+    case publicKey
+    case signature
+    case method
+    case params
+    case code
+    case data
+  }
+  
+  private weak var tab: Tab?
+  
+  init(tab: Tab) {
+    self.tab = tab
+  }
+  
+  static func name() -> String {
+    return "walletSolanaProvider"
+  }
+  
+  func scriptMessageHandlerName() -> String? {
+    return "walletSolanaProvider_\(UserScriptManager.messageHandlerTokenString)"
+  }
+  
+  static func shouldInjectWalletProvider(_ completion: @escaping (Bool) -> Void) {
+    BraveWallet.KeyringServiceFactory.get(privateMode: false)?
+      .keyringInfo(BraveWallet.SolanaKeyringId, completion: { keyring in
+        completion(keyring.isKeyringCreated)
+      })
+  }
+  
+  private struct MessageBody: Decodable {
+    enum Method: String, Decodable {
+      case connect
+      case disconnect
+      case signAndSendTransaction
+      case signMessage
+      case request
+      case signTransaction
+      case signAllTransactions
+    }
+    var securityToken: String
+    var method: Method
+    var args: String?
+    
+    private enum CodingKeys: String, CodingKey {
+      case securityToken = "securitytoken"
+      case method
+      case args
+    }
+  }
+  
+  func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage, replyHandler: @escaping (Any?, String?) -> Void) {
+    Task { @MainActor in
+      guard let tab = tab,
+            !tab.isPrivate,
+            let provider = tab.walletSolProvider,
+            !message.frameInfo.securityOrigin.host.isEmpty, // Fail if there is no last committed URL yet
+            message.frameInfo.isMainFrame, // Fail the request came from 3p origin
+            JSONSerialization.isValidJSONObject(message.body),
+            let messageData = try? JSONSerialization.data(withJSONObject: message.body, options: []),
+            let body = try? JSONDecoder().decode(MessageBody.self, from: messageData),
+            body.securityToken == UserScriptManager.securityTokenString
```

Comment:
What happens if the JSON object passed as the body already contains a property called securityToken?

Reply:
If a `securitytoken` is passed in the JSON object that wasn't added from our `postMessage` (which is randomized every session), it would fail the check here and return and log the ["Failed to handle solana provider communication"](https://github.com/brave/brave-ios/blob/solana-dapps/Client/Wallet/SolanaProviderHelper.swift#L86) error below and not proceed with any solana provider communication.

Reply:
It shouldn't be an issue @kdenhartog because it is passed like:

```javascript
postMessage({"securityToken": OurInjectedToken, message: actualJSObjectContainingSecurityTokenAlready})
```

instead of:
```javascript
postMessage(actualJSObjectContainingSecurityTokenAlready)
```

So the iOS side would receive:
```swift
{
    securityToken: "injectedToken",
    message: {"securityToken": ..., field: "...", otherField: "..."}
}
```

But also, my own PR will merge before this one, and the parsing here will HAVE to change (it cannot use `body.securityToken == UserScriptManager.securityToken` anymore, as there is no such thing.

Reply:
Sounds safe to me, thanks for the context resolving this one.